src/swapRouter.ts

```diff
@@ -300,14 +336,84 @@ export abstract class SwapRouter {
       }
     }
 
-    // refund
-    if (mustRefund) {
+    // must refund when paying in ETH, but with an uncertain input amount
+    if (inputIsNative && sampleTrade.tradeType === TradeType.EXACT_OUTPUT) {
       calldatas.push(Payments.encodeRefundETH())
     }
 
     return {
       calldata: MulticallExtended.encodeMulticall(calldatas, options.deadlineOrPreviousBlockhash),
-      value: toHex(totalValue.quotient),
+      value: toHex(inputIsNative ? totalAmountIn.quotient : ZERO),
+    }
+  }
+
+  /**
+   * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
+   * @param trade to produce call parameters for
+   * @param options options for the call parameters
+   */
+  public static swapAndAddCallParameters(
+    trades:
+      | Trade<Currency, Currency, TradeType>
+      | V2Trade<Currency, Currency, TradeType>
+      | V3Trade<Currency, Currency, TradeType>
+      | (V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>)[],
+    options: SwapOptions,
+    tokenInBalanceDesired: CurrencyAmount<Currency>,
+    tokenOutBalanceDesired: CurrencyAmount<Currency>,
+    position: Position,
+    addLiquidityOptions: AddLiquidityOptions
```

Comment:
per the object at the top, might just consolidate this into options: SwapOptions